Core: Add KLL Datasketch as standard blob types to puffin file

simhadri-g opened the pull request:
…b types to puffin file

Issue: 
https://github.com/apache/iceberg/issues/8198


Hi Everyone,

Hive now supports writing column statistics to puffin files.

The statistics calculated by Hive include histograms, NDV (Number of Distinct Values), Min and Max values, the number of nulls, the number of true values, column name, and column type. You can find the full list of supported stats here: [Link to GitHub](https://github.com/apache/hive/blob/master/ql/src/java/org/apache/hadoop/hive/ql/stats/StatsUtils.java#L854C17-L854C30).

We are updating the description of this PR to request incorporating the KLL datasketch for histograms.

As a result, we are looking to add  KLL datasketch as standard blob types for the puffin file. [Link to GitHub](https://github.com/apache/iceberg/blob/master/core/src/main/java/org/apache/iceberg/puffin/StandardBlobTypes.java)

Any feedback would be greatly appreciated.

Thanks!

simhadri-g committed:
Core: Add KLL Datasketch and Hive ColumnStatisticsObj as standard blob types to puffin file

findepi commented on core/src/main/java/org/apache/iceberg/puffin/StandardBlobTypes.java:
```diff
@@ -26,4 +26,17 @@ private StandardBlobTypes() {}
+
+  /**
+   * A serialized form of <a
+   * href="https://github.com/apache/hive/blob/master/ql/src/java/org/apache/hadoop/hive/ql/stats/StatsUtils.java#L854C17">
+   * HIVE_COLUMN_STATS_OBJ</a>
+   */
+  public static final String HIVE_COLUMN_STATS_OBJ = "column-statistics-obj";
```
Nothing prevents Hive from using arbitrary String IDs for Hive-specific statistics, however I'd encourage to prefix Hive-specific statistics with eg "hive." to avoid conflicts in the future.

If this is supposed to be a "standard blob type", it needs to be documented at https://iceberg.apache.org/puffin-spec/#blob-types in Hive-agnostic way

simhadri-g replied:
Corresponding iceberg-docs PR : https://github.com/apache/iceberg-docs/pull/269

findepi commented on core/src/main/java/org/apache/iceberg/puffin/StandardBlobTypes.java:
```diff
@@ -26,4 +26,17 @@ private StandardBlobTypes() {}
+
+  /**
+   * A serialized form of <a
+   * href="https://github.com/apache/hive/blob/master/ql/src/java/org/apache/hadoop/hive/ql/stats/StatsUtils.java#L854C17">
```
nit: github links to a branch are likely to drift over time or even break (if eg module, package or class are renamed or removed)

simhadri-g replied:
Done : Added a table in the hive wiki for supported column stats and replaced the github link to the wiki link.

findepi commented on core/src/main/java/org/apache/iceberg/puffin/StandardBlobTypes.java:
```diff
@@ -26,4 +26,17 @@ private StandardBlobTypes() {}
+   * A serialized form of KLL sketch produced by the <a
+   * href="https://datasketches.apache.org/">Apache DataSketches</a> library
+   */
+  public static final String APACHE_DATASKETCHES_KLL_SKETCH = "apache_datasketches_kll_sketch";
```
Add documentation to https://iceberg.apache.org/puffin-spec/#blob-types
and follow existing naming convention (hyphenated names).

simhadri-g replied:
Sure. I will try to raise a iceberg-docs PR by tomorrow.

simhadri-g replied:
Corresponding iceberg-docs PR : https://github.com/apache/iceberg-docs/pull/269

simhadri-g commented:
Thanks a lot for the review! :)

simhadri-g committed:
Addressed review comments

simhadri-g commented:
@findepi I have addressed the review comments. 

Can you please have a look at the PR when you are free? 
Thanks in advance!  :)

simhadri-g commented:
Hello @findepi @nastra ,
I would greatly appreciate it if you could find some time to review the pull request.
Thanks!

nastra commented on core/src/main/java/org/apache/iceberg/puffin/StandardBlobTypes.java:
```diff
@@ -26,4 +26,17 @@ private StandardBlobTypes() {}
+   * <a href="https://cwiki.apache.org/confluence/display/Hive/StatsDev#StatsDev-ColumnStatistics">
+   *     Hive columns stats wiki </a>
+   */
+  public static final String HIVE_COLUMN_STATS_OBJ = "column-statistics-obj";
```
should this be named `hive-column-statistics-obj`?

simhadri-g replied:
Yes.
I have updated the name to hive-column-statistics-obj.

nastra commented on core/src/main/java/org/apache/iceberg/puffin/StandardBlobTypes.java:
```diff
@@ -26,4 +26,17 @@ private StandardBlobTypes() {}
   public static final String APACHE_DATASKETCHES_THETA_V1 = "apache-datasketches-theta-v1";
+
+  /**
+   * A serialized form of HIVE_COLUMN_STATS_OBJ. The full list of available stats are provided in the
```
rather than saying `HIVE_COLUMN_STATS_OBJ`, should this say `Hive column stats object`?

simhadri-g replied:
Done, I have updated the same to  "Hive column stats object".

nastra commented on core/src/main/java/org/apache/iceberg/puffin/StandardBlobTypes.java:
```diff
@@ -26,4 +26,17 @@ private StandardBlobTypes() {}
+  public static final String HIVE_COLUMN_STATS_OBJ = "column-statistics-obj";
+
+  /**
+   * A serialized form of KLL sketch produced by the <a
```
```suggestion
   * A serialized form of a KLL sketch produced by the <a
```

simhadri-g replied:
Fixed, thanks!

simhadri-g commented:
Thanks for the review :) 
I will update the PR and get back.

simhadri-g committed:
Addressed review comments and moved the doc update from iceberg-doc to the main iceberg/format

simhadri-g commented:
@nastra , I have updated the PR and moved the iceberg-docs(https://github.com/apache/iceberg-docs/pull/269) changes to the main repo.

Please have a look when you are free. 

Thanks again for the review!

nastra commented on format/puffin-spec.md:
```diff
@@ -126,6 +126,23 @@ The blob metadata for this blob may include following properties:
+
+A serialized form of Hive ColumnStatsObject.
+
+The columnStatsObject supports Histograms, NDV, Min and Max values, Number of nulls, Number of trues, column name, type.
```
```suggestion
The ColumnStatsObject supports Histograms, NDV, Min and Max values, Number of nulls, Number of trues, column name, type.
```

simhadri-g committed:
Update a typo in format/puffin-spec.md

Co-authored-by: Eduard Tudenhoefner <etudenhoefner@gmail.com>

nastra reviewed (approved):
LGTM, @findepi could you review this as well please?

simhadri-g committed:
Fix checkstyle error flagged by :iceberg-core:spotlessJavaCheck

simhadri-g commented:
thanks @nastra for the review! :) 

@findepi Could you please have a look as well? 
Thanks!

ZacBlanco commented:
Any progress update here? It would be great to get the blessing from the Iceberg community for these as supported puffin blob types

bitsondatadev commented:
@findepi @danielcweeks could one of you PTAL this? Don't want this effort to get lost.

simhadri-g commented:
thanks @bitsondatadev !

zhangbutao commented:
Any update? hope we can continue to push this review forward.
Thanks.

tdcmeehan commented on format/puffin-spec.md:
```diff
@@ -126,6 +126,23 @@ The blob metadata for this blob may include following properties:
 
+#### `hive-column-statistics-obj` blob type
+
+A serialized form of Hive ColumnStatsObject.
```
Is what's referenced here the Thrift `ColumnStatisticsObj` in the Hive IDL? https://github.com/apache/hive/blob/ffb1165f59defa66b31b4fd9cb6367b71050071b/standalone-metastore/metastore-common/src/main/thrift/hive_metastore.thrift#L583

If so, I'd recommend correcting the name, and linking to the Thrift IDL, and explicitly calling out that this is Thrift-serialized.

I'm also wondering if we need to think about versioning.  If this is based on the Thrift IDL, I am not sure if those are intended to be persisted.  At the very least, I am concerned if Hive decides to introduce a backwards-incompatible field to this struct, some engines begin to serialize with this newly introduced backwards incompatible field, and other engines begin to attempt to deserialize it with an older IDL, then it will break in the Iceberg library.

Please let me know if I'm misunderstanding anything.

simhadri-g replied:
Hi,
In hive , we writes the statistics to HMS in addition to puffin files.  
So the Thrift ColumnStatisticsObj is used to write the statistics to HMS.

hive-column-statistics-obj referred here is a serialized using the org.apache.commons.lang3.SerializationUtils and stored into puffin file.

simhadri-g commented:
Hi everyone, 
I would be most grateful if we could get help with reviewing this. 
Thanks!

deniskuzZ commented:
hi @simhadri-g, would you mind reopening this PR?

simhadri-g commented:
Hi Denys!
I think i don't have permission to reopen a closed PR in iceberg. Should i raise a new PR?

deniskuzZ commented:
> Hi Denys! I think , I don't have the github permissions to reopen a closed PR in iceberg repo. Should i raise a new PR?

hey @simhadri-g, if you'd like to pursue this I would recommend opening a new one, but split into 2 parts: Hive ColumnStatistics and KLL, Datasketch

deniskuzZ commented:
thanks @nastra! I would really help if we could get this in.

simhadri-g commented:
thanks @nastra and @deniskuzZ  !  

I will resolve the merge conflicts and update this PR.  :)

nastra commented:
@findepi Could you please have a look as well?

deniskuzZ commented on format/puffin-spec.md:
```diff
@@ -181,6 +181,23 @@ for Puffin v1.
 [roaring-bitmap-portable-serialization]: https://github.com/RoaringBitmap/RoaringFormatSpec?tab=readme-ov-file#extension-for-64-bit-implementations
 [roaring-bitmap-general-layout]: https://github.com/RoaringBitmap/RoaringFormatSpec?tab=readme-ov-file#general-layout
 
+#### `hive-column-statistics-obj` blob type
```
should I start [DISCUSS] thread first or go ahead with [VOTE]?

nastra replied:
you can also start with a DISCUSS thread first in order to have a short discussion on the introduced changes

rdblue commented on format/puffin-spec.md:
```diff
@@ -181,6 +181,23 @@ for Puffin v1.
+
+The ColumnStatsObject supports Histograms, NDV, Min and Max values, Number of nulls, Number of trues, column name, type.
+A full list of supported statistics is listed in the table here:
+[ColumnStatistics](https://cwiki.apache.org/confluence/display/Hive/StatsDev#StatsDev-ColumnStatistics)
```
I don't see much value in adding this.

NDV sketches are already supported using Theta sketches so this would duplicate the purpose of an existing sketch. Is there sufficient value to justify the difference? This doesn't provide enough context to tell.

In addition, the Iceberg manifest format already covers value count, lower bounds, upper bounds, number of nulls, and number of NaNs. The partition statistics files provide a way to aggregate those beyond the file level, and we use snapshot summaries for table-level stats. I'm not sure what value this would provide.

deniskuzZ replied:
hi @rdblue,
thanks for checking this PR!
 
`The partition statistics files provide a way to aggregate those beyond the file level` 
Does iceberg provide build-in support to get an aggregated Column stats? I mean, is there some library/service that generates partition files with an aggregated column stats? 
AFAIK we only do this for basic stats : https://github.com/apache/iceberg/pull/11216
 
If yes, could you please point me to the code where is that done? I had an impression that from colstats only NDV is calculated and stored in partition files.

How about:
1. bitvectors - used to improve stats estimations for IN operator
2. histogram - histogram statistics, which are particularly useful for skewed data and range predicates (KLL data sketches)

deniskuzZ replied:
hi @rdblue,
I'm sorry, it seems we had pretty limited knowledge in that area and now I think we finally get your point.
I've drafted a small doc with the proposal and our intent: https://docs.google.com/document/d/11Rp-irqb4L4Qpdxr6l83bA4IRsfw3AAyR8wokNe1r80/edit?usp=sharing
Could you please take a quick look and suggest if that is a valid proposal. 
Thank you!

rdblue commented on format/puffin-spec.md:
```diff
@@ -181,6 +181,23 @@ for Puffin v1.
+Apache-Datasketches-KLL-sketch is an implementation of a very compact quantiles
+sketch with lazy compaction scheme and nearly optimal accuracy per bit.
+
+Histograms are derived from this sketch.
```
This isn't much to go on. How is the sketch calculated? Does this support a single column or multiple columns?

deniskuzZ replied:
more details: https://issues.apache.org/jira/browse/HIVE-26221

danielcweeks commented:
I would agree with @rdblue's comment that there's a lot of stats overlap with what exists elsewhere and I'm not convinced this serialized format is standardized well enough to be used easily outside of the Hive/Impala projects.

Could we identify the specific stats that we feel are valuable and focus on those independently of what is captured elsewhere?

deniskuzZ committed:
removed Hive ColumnStatsObject

deniskuzZ commented:
hi @nastra, could you please help push it forward? I've updated the PR to include the KLL sketch only.

nastra commented:
@deniskuzZ this is a spec change and needs to go through a DISCUSS & VOTE thread on the mailing list

deniskuzZ commented:
> @deniskuzZ this is a spec change and needs to go through a DISCUSS & VOTE thread on the mailing list

This was a thread: https://lists.apache.org/thread/ws63loz1snsrwtdl7f9yqgr16vncx6w0
I don't know what else I should do other than just abandon the PR. 
It looks like this change generated some interest until the recent comments, which I hope have been addressed, but I’m not sure since I haven't received any feedback.

nastra commented:
@deniskuzZ you might just want to follow-up on the dev list to revive the discussion. Once it's clear in which direction the proposal is going and what Spec changes are required, you'd go ahead and create a VOTE thread for the Spec changes (this PR) where people can vote on the changes

thomasrebele commented:
There are alternative histogram sketches to KLL, with various pros and cons. Would it make sense to make this PR a bit more general, so that the type of the sketch is not hardcoded? It would be nice if the engine could decide which histogram algorithm it wants to use.

deniskuzZ commented:
> There are alternative histogram sketches to KLL, with various pros and cons. Would it make sense to make this PR a bit more general, so that the type of the sketch is not hardcoded? It would be nice if the engine could decide which histogram algorithm it wants to use.

@thomasrebele, apache-datasketches-kll-v1 is itself a sketch type. If the sketch type isn't encoded, how would we determine what algorithm the sketch represents? 
Could you elaborate on what you have in mind by saying make it more generic?

thomasrebele commented:
Apache Datasketches allows to create ReqSketch or t-digest histogram sketches. Other histogram sketch algorithms might be added in the future. I mean that it would be nice to be able to store a t-digest sketch in Iceberg, instead of a KLL sketch. If we follow that approach, it should be possible to define the type per column, so that, e.g., column `a`  could have a KLL sketch, and column `b` a t-digest sketch.

deniskuzZ commented:
> Apache Datasketches allows to create ReqSketch or t-digest histogram sketches. Other histogram sketch algorithms might be added in the future. I mean that it would be nice to be able to store a t-digest sketch in Iceberg, instead of a KLL sketch. If we follow that approach, it should be possible to define the type per column, so that, e.g., column `a` could have a KLL sketch, and column `b` a t-digest sketch.

Is this what you mean:
````
sketch {
  type: KLL | TDIGEST | REQ
  payload: bytes
}
````
